src/flb_async_timer.c

```diff
@@ -0,0 +1,169 @@
+    struct mk_list *tmp;
+    struct mk_list *head;
+    mk_list_foreach_safe(head, tmp, &config->outputs) {
+        o_ins = mk_list_entry(head, struct flb_output_instance, _head);
+        flb_async_timer_cleanup(o_ins->async_timer_list_destroy);
+    }
+}
+
+int flb_sched_out_async_timer_cb_create(struct flb_sched *sched, int type, int ms,
+                                        struct flb_output_instance *o_ins,
+                                        char *job_name,
+                                        void (*async_cb)(struct flb_config *, void *),
+                                        void *data, struct flb_sched_timer **out_timer)
+{
+    struct flb_out_async_timer_cb_data *timer_data;
+
+    timer_data = flb_calloc(1, sizeof(struct flb_out_async_timer_cb_data));
+    if (!timer_data) {
+        return;
+    }
+
+    timer_data->ins = o_ins;
+    timer_data->job_name = job_name;
+    timer_data->cb = async_cb;
+    timer_data->data = data;
+
+    return flb_sched_timer_cb_create(sched, type, ms, flb_out_async_sched_timer_cb, timer_data, NULL);
+}
+
+/* Used in engine flb_running_count */
+int flb_async_timers_size(struct flb_output_instance *ins)
+{
+    int size = 0;
+
+    if (flb_output_is_threaded(ins) == FLB_TRUE) {
+        /*
+         * On threaded mode, we need to count the active co-routines of
+         * every running thread of the thread pool.
+         */
+        size = flb_thread_pool_async_timers_size(ins);
+    }
+    else {
+        size = mk_list_size(&ins->async_timer_list);
+    }
+
+    return size;
+}
+
+void flb_async_timers_print(struct mk_list *async_timer_list)
+{ 
+    struct flb_out_async_timer *async_timer;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    int n = mk_list_size(async_timer_list);
+    if (n != 0) {
+        /* get one coro for the job_name */
+        mk_list_foreach_safe(head, tmp, async_timer_list) {
+            async_timer = mk_list_entry(head, struct flb_out_async_timer, _head);
+            if (async_timer != NULL) {
+                flb_info("[task]   output=%s still running %d %s(s)",
+                         async_timer->o_ins->alias, n, async_timer->timer_data->job_name);
+                break;
+            }
+        }
+    }
+}
+
+/* Used in engine flb_running_print */
+void flb_out_async_timers_print(struct flb_output_instance *ins)
+{
+    if (flb_output_is_threaded(ins) == FLB_TRUE) {
+        flb_thread_pool_async_timers_print(ins);
+    }
+    else {
+        flb_async_timers_print(&ins->async_timer_list);
+    }
+}
+
+int flb_thread_pool_async_timers_size(struct flb_output_instance *ins)
+{
+    int n;
+    int size = 0;
+    struct mk_list *head;
+    struct flb_tp *tp = ins->tp;
+    struct flb_tp_thread *th;
+    struct flb_out_thread_instance *th_ins;
+
+    mk_list_foreach(head, &tp->list_threads) {
+        th = mk_list_entry(head, struct flb_tp_thread, _head);
+        if (th->status != FLB_THREAD_POOL_RUNNING) {
```

Comment:
One interesting thing, `FLB_THREAD_POOL_RUNNING` just tells you the thread has been started, the `FLB_THREAD_POOL_STOPPED` value is never set.